File: capacitor/plugins-library/android/src/main/java/com/krikoo/capacitor/DataStorageUtils.java
package com.krikoo.capacitor;

import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import org.json.JSONObject;

public class DataStorageUtils {

  public static void error(String message, SqliteDB db, PluginCall call) {
    db.close();
    call.reject(message);
  }

  public static String getNoEmptyString(String key, PluginCall call) {
    String table = call.getString(key);
    return table == null || table.isEmpty() ? null : table;
  }

  public static JSObject getObjectToStore(String key, PluginCall call) {
    JSObject value = new JSObject();

    String textValue = call.getString(key, null);
    if (textValue != null) {
      value.put("value", textValue);
    }

    Integer intValue = call.getInt(key, null);
    if (intValue != null) {
      value.put("value", intValue);
    }

    JSArray arrayValue = call.getArray(key, null);
    if (arrayValue != null) {
      value.put("value", arrayValue);
    }

    JSObject objectValue = call.getObject(key, null);
    if (objectValue != null) {
      value.put("value", objectValue);
    }

    return value.length() == 0 ? null : value;
  }

  public static JSObject jsonParse(String text) {
    try {
      return new JSObject(text);
    } catch (Exception e) {
      Log.i("DATA STORAGE", String.format("Bad JSON: %s", text));
      return null;
    }
  }

  public static void success(JSObject result, SqliteDB db, PluginCall call) {
    db.close();
    call.success(result);
  }

}

File: capacitor/plugins-library/ios/Plugin/DataStorage/DataStorageUtils.swift
import Foundation
import Capacitor

class DataStorageUtils {
    
    public static let DATA_STORAGE_ERROR_HASH = "KRIKOO_ERROR_HASH"
    
    public static func error(_ message: String, _ db: SqliteDB, _ call: CAPPluginCall) {
        db.close()
        call.reject(message)
    }
    
    public static func getNoEmptyString(_ key: String, _ call: CAPPluginCall) -> String? {
        guard let value = call.getString(key) else {
            return nil
        }
        return value.isEmpty ? nil : value
    }
    
    public static func getObjectToStore(_ key: String, _ call: CAPPluginCall) -> [String:Any]? {
        // TODO: Add type to avoid problems with boolean and numbers
        // TODO: Now transforms 0 to false.
        
        if let value = call.getBool(key, nil) {
            return ["value": value]
        }
        
        if let value = call.getDouble(key, nil) {
            return ["value": value]
        }
        
        if let value = call.getString(key, nil) {
            return ["value": value]
        }
        
        if let value = call.getArray(key, Any.self, nil) {
            return ["value": value]
        }
        
        if let value = call.getObject(key, defaultValue: nil) {
            return ["value": value]
        }
        
        return nil
    }
    
    public static func jsonStringify(json: [String:Any], prettyPrinted: Bool = false) -> String? {
        var options: JSONSerialization.WritingOptions = []
        if prettyPrinted {
            options = JSONSerialization.WritingOptions.prettyPrinted
        }
        
        do {
            let data = try JSONSerialization.data(withJSONObject: json, options: options)
            if let string = String(data: data, encoding: String.Encoding.utf8) {
                return string
            }
        } catch {
            print("DATA STORAGE -> Stringify Error: \(error)")
            return nil
        }
        return nil
    }
    
    public static func jsonParse(_ text: String) -> [String:Any]? {
        let data = text.data(using: .utf8)!
        do {
            if let jsonArray = try JSONSerialization.jsonObject(with: data, options : .allowFragments) as? [String:Any] {
                return jsonArray
            } else {
                print("DATA STORAGE -> Bad JSON: \(text)")
                return nil
            }
        } catch {
            print("DATA STORAGE -> JSON Error: \(error)")
            return nil
        }
    }
    
    public static func success(_ result: [String:Any], _ db: SqliteDB, _ call: CAPPluginCall) {
        db.close()
        call.success(result)
    }
    
}
